Add code and fill readme (#1)
jamesTait-jt/go-performance-tips: README.md

    }

    return s[:indexOfNext], s[indexOfNext + 1:]
}
```

Now we can use this in our person parsing and hopefully we'll see an improvement in performance

```go
func PersonEfficient(s string) person {
    first, last := ParseUntil(s, ' ')

    n := person{
        firstName: first,
        lastName: last,
    }

    return n
}
```

As you can see, we're removed the slice, so now, nothing should escape to the heap. Let's benchmark it to be sure

```go
func Benchmark_PersonEfficient(b *testing.B) {
    s := "joe bloggs"

    b.ReportAllocs()
    for i := 0 ; i < b.N ; i++ {
        _ = PersonEfficient(s)
    }
}
```

Let's run both benchmarks together so we can compare

```sh
Benchmark_Person
Benchmark_Person-10               32530510                36.55 ns/op           32 B/op          1 allocs/op
Benchmark_PersonEfficient
Benchmark_PersonEfficient-10      261291204                4.577 ns/op           0 B/op          0 allocs/op
```

As we can see, the allocation has gone and the function is also nearly 10x faster! So, not only have we saved on the garbage collection time, but by avoiding writing to the heap we have saved ourselves an expensive operation.

> This method is be very useful when we want to parse a message from another service and convert it into a structure that is known by our service. I have shown it with strings, but the same method can be used with any slice, like a slice of bytes, for example.

Reviewer: ```suggestion
This method is very useful when we want to parse a message from another service and convert it into a structure that is known by our service. I have shown it with strings, but the same method can be used with any slice, like a slice of bytes, for example.
```

```go
func PersonBytes(bs []byte) (person, error) {
    first, rest := ParseUntilBytes(bs, '\x1f')
    last, rest := ParseUntilBytes(rest, '\x1f')
    ageBytes, rest := ParseUntilBytes(rest, '\x1f')
    heightBytes, _ := ParseUntilBytes(rest, '\x1f')

    ageStr := string(ageBytes)
    age, err := strconv.Atoi(ageStr)
    if err != nil {
        return person{}, err
    }

    heightStr := string(heightBytes)
    height, err := strconv.ParseFloat(heightStr, 64)
    if err != nil {
        return person{}, err
    }

    p := person{
        firstName: string(first),
        lastName: string(last),
        age: age,
        height: height,
    }

    return p, nil
}
```

And updating our benchmarking function we get

```go
func Benchmark_PersonBytes(b *testing.B) {
    bs := []byte("joe\x1fbloggs\x1f25\x1f182.3")

    b.ReportAllocs()
    for i := 0 ; i < b.N ; i++ {
        PersonBytes(bs)
    }
}
```

```sh
Benchmark_PersonBytes
Benchmark_PersonBytes-10        12817989                91.76 ns/op           16 B/op          4 allocs/op
```

> So we are allocating even more memory to the heap due to the string conversions of the `age` and `height` fields. Now, let's introduce some unsafe functions to help us out (these ones are avtually pretty safe!)

Reviewer: ```suggestion
So we are allocating even more memory to the heap due to the string conversions of the `age` and `height` fields. Now, let's introduce some unsafe functions to help us out (these ones are actually pretty safe!)
```

As we have discussed, allocating memory at runtime is expensive. So if we know at compile time the upper bound of a slice's size, it is worth pre-allocating the capacity with a constant. Here is an example of a growing slice without pre allocating memory

```go
func growSlice() {
    toAdd := []byte("123456")
    bs := []byte{}

    bs = append(bs, toAdd...)
    bs = append(bs, toAdd...)
    bs = append(bs, toAdd...)
}
```

Looking at the benchmarks it is easy to see that a lot of heap allocations are taking place.

```sh
Benchmark_GrowSlice
Benchmark_GrowSlice-10                  23496968                48.22 ns/op           56 B/op          3 allocs/op
```

every time the slice grows past its capacity, the entire slice needs to be reallocated on the heap with a larger capacity. The method is useful in a large number of use cases where we do not know the capacity of the slice, however, if we do know it then we can do a lot better and save on those costly runtime allocations.

```go
func growSlicePreAllocate() {
    toAdd := []byte("123456")
    bs := make([]byte, 0, 18)

    bs = append(bs, toAdd...)
    bs = append(bs, toAdd...)
    bs = append(bs, toAdd...)
}
```

> Here, we know that the `toAdd` bytes will be appended to the `bs` slice 3 times, giving us a final slice of length `6 * 3 = 18` bytes. So we pre-allocate this much memory which allows the compiler to allocate memory on the stack rather than allocating on the heap at runtime, and then having to handle the garbage colection as well.

Reviewer: ```suggestion
Here, we know that the `toAdd` bytes will be appended to the `bs` slice 3 times, giving us a final slice of length `6 * 3 = 18` bytes. So we pre-allocate this much memory which allows the compiler to allocate memory on the stack rather than allocating on the heap at runtime, and then having to handle the garbage collection as well.
```

The reason is that the compiler has no knowledge about the implementation of an interface method at compile time. So it is possible for the implementation to store the value somewhere (maybe a package level variable) and the compiler will have no idea.

Let's say this is the case, and the method argument was a pointer. So the package level variable is pointing to the same location as the original pointer that we passes into the method. However, When the original pointer goes out of scope, the value pointed to by the pointer will be trashed. And thus, the package level variable will be pointing to garbage. Because of this, the pointer value is moved to the heap so that it is not trashed when the stackframe is popped.

Here is an example with some benchmarks

```go
package interfaces

type NoOpper interface {
    Int(n int)
    IntPtr(n *int)
}

type nothing struct {}

func (no nothing) Int(n int) {}

func (no nothing) IntPtr(n *int) {}

//go:noinline
func NoOpInt(no NoOpper, n int) {
    no.Int(n)
}

//go:noinline
func NoOpIntPtr(no NoOpper, n int) {
    no.IntPtr(&n)
}
```

> Here we have two trivial interface methods, one accepts an `int` and the other accepts an `*int`. We use `//go:noinline` to prevent the compiler from optimising out the function call as we can't rely on this in a producstion setting with ever-changing code. 

Reviewer: ```suggestion
Here we have two trivial interface methods, one accepts an `int` and the other accepts an `*int`. We use `//go:noinline` to prevent the compiler from optimising out the function call as we can't rely on this in a production setting with ever-changing code. 
```

Even though these methods don't do anything, the expectation is that with the `*int` version, `n` will be allocated to the heap.

```go
package interfaces

import (
    "testing"
)

func BenchmarkNoOpInt(b *testing.B) {
    b.ReportAllocs()

    no := nothing{}
    n := 10

    for i := 0 ; i < b.N ; i++ {
        NoOpInt(no, n)
    }
}

func BenchmarkNoOpIntPtr(b *testing.B) {
    b.ReportAllocs()

    no := nothing{}
    n := 10

    for i := 0 ; i < b.N ; i++ {
        NoOpIntPtr(no, n)
    }
}
```

```sh
BenchmarkNoOpInt
BenchmarkNoOpInt-10             578409886                2.068 ns/op           0 B/op          0 allocs/op
BenchmarkNoOpIntPtr
BenchmarkNoOpIntPtr-10          131864689                9.051 ns/op           8 B/op          1 allocs/op
```

> As expected, the variable is allocated to the heap. Of course, the tradeoff is that if we did not pass a pointer, we would have to take a copy of the parameter to pass it into the called function's stackframe. The reccomendation here is to benchmark both options and see which is better for your application.

Reviewer: ```suggestion
As expected, the variable is allocated to the heap. Of course, the trade-off is that if we did not pass a pointer, we would have to take a copy of the parameter to pass it into the called function's stackframe. The recommendation here is to benchmark both options and see which is better for your application.
```
